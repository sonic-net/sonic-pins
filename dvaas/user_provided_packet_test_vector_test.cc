// Empowers users to specify custom packet test vectors that can be validated
// by DVaaS or Arriba.

// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "dvaas/user_provided_packet_test_vector.h"

#include <iostream>
#include <string>
#include <vector>

#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/string_view.h"
#include "dvaas/test_vector.h"
#include "dvaas/test_vector.pb.h"
#include "gmock/gmock.h"
#include "gtest/gtest.h"
#include "gutil/collections.h"
#include "gutil/proto.h"
#include "gutil/status_matchers.h"
#include "gutil/testing.h"
#include "p4_pdpi/packetlib/packetlib.pb.h"
#include "p4_pdpi/testing/test_p4info.h"

namespace dvaas {
namespace {

using ::gutil::IsOk;
using ::gutil::PrintTextProto;
using ::testing::Not;

struct TestCase {
  std::string description;
  std::vector<PacketTestVector> vectors;
};

void RunTestCase(const TestCase& test_case) {
  // Print header.
  std::cout << std::string(80, '=') << "\n"
            << "InternalizeUserProvidedTestVectors Test: "
            << test_case.description << "\n"
            << std::string(80, '=') << "\n";

  // Print input.
  std::cout << "-- Input ---------------------------------------------------\n";
  for (int i = 0; i < test_case.vectors.size(); ++i) {
    std::cout << "-- Input Packet Test Vector #" << (i + 1) << " --\n"
              << PrintTextProto(test_case.vectors[i]);
  }
  std::cout << "\n";

  // Print output.
  std::cout << "-- Output --------------------------------------------------\n";
  pdpi::IrP4Info ir_p4info = pdpi::GetTestIrP4Info();
  absl::StatusOr<PacketTestVectorById> output =
      LegitimizeUserProvidedTestVectors(test_case.vectors, ir_p4info);
  if (!output.ok()) {
    // Print error without stack trace, for golden testing.
    std::cout << "ERROR: "
              << output.status().ToString(
                     absl::StatusToStringMode::kWithNoExtraData)
              << "\n\n";
    return;
  }

  // Print output vectors in order of input vectors.
  for (int i = 0; i < test_case.vectors.size(); ++i) {
    const PacketTestVector& original_vector = test_case.vectors[i];
    ASSERT_OK_AND_ASSIGN(
        int tag,
        ExtractTestPacketTag(original_vector.input().packet().parsed()));
    ASSERT_OK_AND_ASSIGN(const PacketTestVector* internalized_vector,
                         gutil::FindPtrOrStatus(*output, tag));
    ASSERT_OK_AND_ASSIGN(
        std::string diff,
        gutil::ProtoDiff(original_vector, *internalized_vector));
    std::cout << "-- Internalized Packet Test Vector #" << (i + 1) << " --\n"
              << "test packet ID extracted from payload: " << tag << "\n"
              << "diff of internalized vector vs input vector:\n"
              << diff << "\n";
  }
  std::cout << "\n";
}

std::vector<TestCase> GetPositiveTestCases() {
  std::vector<TestCase> test_cases;

  test_cases.emplace_back() = TestCase{
      .description = "empty list of vectors",
      .vectors = {},
  };

  test_cases.emplace_back() = TestCase{
      .description = "single Ethernet packet expected to be dropped",
      .vectors =
          {
              gutil::ParseProtoOrDie<PacketTestVector>(R"pb(
                input {
                  type: DATAPLANE
                  packet {
                    port: "1"
                    parsed {
                      headers {
                        ethernet_header {
                          ethernet_destination: "ff:ee:dd:cc:bb:aa"
                          ethernet_source: "55:44:33:22:11:00"
                          ethertype: "0x000f"
                        }
                      }
                      payload: "test packet #42"
                    }
                  }
                }
                acceptable_outputs {}
              )pb"),
          },
  };

  test_cases.emplace_back() = TestCase{
      .description = "single IPv6 packet expected to be forwarded",
      .vectors =
          {
              gutil::ParseProtoOrDie<PacketTestVector>(R"pb(
                input {
                  type: DATAPLANE
                  packet {
                    port: "1"
                    parsed {
                      headers {
                        ethernet_header {
                          ethernet_destination: "ff:ee:dd:cc:bb:aa"
                          ethernet_source: "55:44:33:22:11:00"
                          ethertype: "0x86dd"
                        }
                      }
                      headers {
                        ipv6_header {
                          version: "0x6"
                          dscp: "0x1b"
                          ecn: "0x1"
                          flow_label: "0x12345"
                          next_header: "0xfd"
                          hop_limit: "0xf2"
                          ipv6_source: "2001::2"
                          ipv6_destination: "2001::4"
                        }
                      }
                      payload: "test packet #42"
                    }
                  }
                }
                acceptable_outputs {
                  packets {
                    port: "8"
                    parsed {
                      headers {
                        ethernet_header {
                          ethernet_destination: "1:2:3:4:5:6"
                          ethernet_source: "6:5:4:3:2:1"
                          ethertype: "0x86dd"
                        }
                      }
                      headers {
                        ipv6_header {
                          version: "0x6"
                          dscp: "0x1b"
                          ecn: "0x1"
                          flow_label: "0x12345"
                          next_header: "0xfd"
                          hop_limit: "0xf1"
                          ipv6_source: "2001::2"
                          ipv6_destination: "2001::4"
                        }
                      }
                      payload: "test packet #42"
                    }
                  }
                }
              )pb"),
          },
  };

  test_cases.emplace_back() = TestCase{
      .description = "several packets with different IDs",
      .vectors =
          {
              gutil::ParseProtoOrDie<PacketTestVector>(R"pb(
                input {
                  type: DATAPLANE
                  packet {
                    port: "1"
                    parsed {
                      headers {
                        ethernet_header {
                          ethernet_destination: "42:42:42:42:42:42"
                          ethernet_source: "42:42:42:42:42:42"
                          ethertype: "0x000f"
                        }
                      }
                      payload: "test packet #42"
                    }
                  }
                }
                acceptable_outputs {}
              )pb"),
              gutil::ParseProtoOrDie<PacketTestVector>(R"pb(
                input {
                  type: DATAPLANE
                  packet {
                    port: "1"
                    parsed {
                      headers {
                        ethernet_header {
                          ethernet_destination: "5:5:5:5:5:5"
                          ethernet_source: "5:5:5:5:5:5"
                          ethertype: "0x000e"
                        }
                      }
                      payload: "test packet #5"
                    }
                  }
                }
                acceptable_outputs {}
              )pb"),
          },
  };

  test_cases.emplace_back() = TestCase{
      .description = "packet_in with matching tag",
      .vectors =
          {
              gutil::ParseProtoOrDie<PacketTestVector>(R"pb(
                input {
                  type: DATAPLANE
                  packet {
                    port: "1"
                    parsed {
                      headers {
                        ethernet_header {
                          ethernet_destination: "ff:ee:dd:cc:bb:aa"
                          ethernet_source: "55:44:33:22:11:00"
                          ethertype: "0x000f"
                        }
                      }
                      payload: "test packet #42"
                    }
                  }
                }
                acceptable_outputs {
                  packet_ins {
                    metadata {
                      name: "ingress_port"
                      value: { hex_str: "0x00f" }
                    }
                    metadata {
                      name: "target_egress_port"
                      value: { str: "1" }
                    }
                    parsed {
                      headers {
                        ethernet_header {
                          ethernet_destination: "5:5:5:5:5:5"
                          ethernet_source: "5:5:5:5:5:5"
                          ethertype: "0x000f"
                        }
                      }
                      payload: "test packet #42"
                    }
                  }
                }
              )pb"),
          },
  };
  return test_cases;
}  // namespace

std::vector<TestCase> GetNegativeTestCases() {
  std::vector<TestCase> test_cases;

  test_cases.emplace_back() = TestCase{
      .description = "input type != DATAPLANE",
      .vectors =
          {
              gutil::ParseProtoOrDie<PacketTestVector>(R"pb(
                input {
                  type: SUBMIT_TO_INGRESS
                  packet {
                    port: "1"
                    parsed {
                      headers {
                        ethernet_header {
                          ethernet_destination: "ff:ee:dd:cc:bb:aa"
                          ethernet_source: "55:44:33:22:11:00"
                          ethertype: "0x000f"
                        }
                      }
                      payload: "test packet #42"
                    }
                  }
                }
                acceptable_outputs {}
              )pb"),
          },
  };

  test_cases.emplace_back() = TestCase{
      .description = "missing expectation",
      .vectors =
          {
              gutil::ParseProtoOrDie<PacketTestVector>(R"pb(
                input {
                  type: DATAPLANE
                  packet {
                    port: "1"
                    parsed {
                      headers {
                        ethernet_header {
                          ethernet_destination: "42:42:42:42:42:42"
                          ethernet_source: "42:42:42:42:42:42"
                          ethertype: "0x000f"
                        }
                      }
                      payload: "test packet #42"
                    }
                  }
                }
              )pb"),
          },
  };

  test_cases.emplace_back() = TestCase{
      .description = "missing test packet ID",
      .vectors =
          {
              gutil::ParseProtoOrDie<PacketTestVector>(R"pb(
                input {
                  type: DATAPLANE
                  packet {
                    port: "1"
                    parsed {
                      headers {
                        ethernet_header {
                          ethernet_destination: "42:42:42:42:42:42"
                          ethernet_source: "42:42:42:42:42:42"
                          ethertype: "0x0002"
                        }
                      }
                      payload: "Hi"
                    }
                  }
                }
              )pb"),
          },
  };

  test_cases.emplace_back() = TestCase{
      .description = "inconsistent test packet ID in input vs output",
      .vectors =
          {
              gutil::ParseProtoOrDie<PacketTestVector>(R"pb(
                input {
                  type: DATAPLANE
                  packet {
                    port: "1"
                    parsed {
                      headers {
                        ethernet_header {
                          ethernet_destination: "42:42:42:42:42:42"
                          ethernet_source: "42:42:42:42:42:42"
                          ethertype: "0x000f"
                        }
                      }
                      payload: "test packet #42"
                    }
                  }
                }
                acceptable_outputs {
                  packets {
                    port: "8"
                    parsed {
                      headers {
                        ethernet_header {
                          ethernet_destination: "5:5:5:5:5:5"
                          ethernet_source: "5:5:5:5:5:5"
                          ethertype: "0x000f"
                        }
                      }
                      payload: "test packet #24"
                    }
                  }
                }
              )pb"),
          },
  };

  test_cases.emplace_back() = TestCase{
      .description = "several packets with same test packet ID",
      .vectors =
          {
              gutil::ParseProtoOrDie<PacketTestVector>(R"pb(
                input {
                  type: DATAPLANE
                  packet {
                    port: "1"
                    parsed {
                      headers {
                        ethernet_header {
                          ethernet_destination: "42:42:42:42:42:42"
                          ethernet_source: "42:42:42:42:42:42"
                          ethertype: "0x000f"
                        }
                      }
                      payload: "test packet #42"
                    }
                  }
                }
                acceptable_outputs {}
              )pb"),
              gutil::ParseProtoOrDie<PacketTestVector>(R"pb(
                input {
                  type: DATAPLANE
                  packet {
                    port: "1"
                    parsed {
                      headers {
                        ethernet_header {
                          ethernet_destination: "5:5:5:5:5:5"
                          ethernet_source: "5:5:5:5:5:5"
                          ethertype: "0x000f"
                        }
                      }
                      payload: "test packet #42"
                    }
                  }
                }
                acceptable_outputs {}
              )pb"),
          },
  };

  test_cases.emplace_back() = TestCase{
      .description = "invalid input packet",
      .vectors =
          {
              gutil::ParseProtoOrDie<PacketTestVector>(R"pb(
                input {
                  type: DATAPLANE
                  packet {
                    port: "1"
                    parsed {
                      headers {
                        ethernet_header {
                          ethernet_destination: "5:5:5:5:5:5"
                          ethernet_source: "5:5:5:5:5:5"
                          ethertype: "0x86dd"  # IPv6.
                        }
                      }
                      # Missing IPv6 header.
                      payload: "test packet #42"
                    }
                  }
                }
                acceptable_outputs {}
              )pb"),
          },
  };

  test_cases.emplace_back() = TestCase{
      .description = "packet_in with mismatching tag",
      .vectors =
          {
              gutil::ParseProtoOrDie<PacketTestVector>(R"pb(
                input {
                  type: DATAPLANE
                  packet {
                    port: "1"
                    parsed {
                      headers {
                        ethernet_header {
                          ethernet_destination: "ff:ee:dd:cc:bb:aa"
                          ethernet_source: "55:44:33:22:11:00"
                          ethertype: "0x000f"
                        }
                      }
                      payload: "test packet #42"
                    }
                  }
                }
                acceptable_outputs {
                  packet_ins {
                    metadata {
                      name: "ingress_port"
                      value: { hex_str: "0x00f" }
                    }
                    metadata {
                      name: "target_egress_port"
                      value: { str: "1" }
                    }
                    parsed {
                      headers {
                        ethernet_header {
                          ethernet_destination: "5:5:5:5:5:5"
                          ethernet_source: "5:5:5:5:5:5"
                          ethertype: "0x000f"
                        }
                      }
                      payload: "test packet #24"
                    }
                  }
                }
              )pb"),
          },
  };

  test_cases.emplace_back() = TestCase{
      .description = "packet_in with malformed metadata",
      .vectors =
          {
              gutil::ParseProtoOrDie<PacketTestVector>(R"pb(
                input {
                  type: DATAPLANE
                  packet {
                    port: "1"
                    parsed {
                      headers {
                        ethernet_header {
                          ethernet_destination: "ff:ee:dd:cc:bb:aa"
                          ethernet_source: "55:44:33:22:11:00"
                          ethertype: "0x000f"
                        }
                      }
                      payload: "test packet #42"
                    }
                  }
                }
                acceptable_outputs {
                  packet_ins {
                    metadata {
                      name: "foobar"
                      value: { str: "foobar" }
                    }
                    metadata {
                      name: "ingress_port"
                      value: { hex_str: "0x00f" }
                    }
                    metadata {
                      name: "target_egress_port"
                      value: { str: "1" }
                    }
                    parsed {
                      headers {
                        ethernet_header {
                          ethernet_destination: "5:5:5:5:5:5"
                          ethernet_source: "5:5:5:5:5:5"
                          ethertype: "0x000f"
                        }
                      }
                      payload: "test packet #24"
                    }
                  }
                }
              )pb"),
          },
  };
  return test_cases;
}

TEST(InternalizeUserProvidedTestVectorsTest, PositiveTestCases) {
  pdpi::IrP4Info ir_p4info = pdpi::GetTestIrP4Info();
  for (TestCase test_case : GetPositiveTestCases()) {
    EXPECT_THAT(LegitimizeUserProvidedTestVectors(test_case.vectors, ir_p4info),
                IsOk());
    RunTestCase(test_case);
  }
}

TEST(InternalizeUserProvidedTestVectorsTest, NegativeTestCases) {
  pdpi::IrP4Info ir_p4info = pdpi::GetTestIrP4Info();
  for (TestCase test_case : GetNegativeTestCases()) {
    EXPECT_THAT(LegitimizeUserProvidedTestVectors(test_case.vectors, ir_p4info),
                Not(IsOk()));
    RunTestCase(test_case);
  }
}

}  // namespace
}  // namespace dvaas
