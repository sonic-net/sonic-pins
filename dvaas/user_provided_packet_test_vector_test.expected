================================================================================
InternalizeUserProvidedTestVectors Test: empty list of vectors
================================================================================
-- Input ---------------------------------------------------

-- Output --------------------------------------------------

================================================================================
InternalizeUserProvidedTestVectors Test: single Ethernet packet expected to be dropped
================================================================================
-- Input ---------------------------------------------------
-- Input Packet Test Vector #1 --
input {
  type: DATAPLANE
  packet {
    port: "1"
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "ff:ee:dd:cc:bb:aa"
          ethernet_source: "55:44:33:22:11:00"
          ethertype: "0x000f"
        }
      }
      payload: "test packet #42"
    }
  }
}
acceptable_outputs {
}

-- Output --------------------------------------------------
-- Internalized Packet Test Vector #1 --
test packet ID extracted from payload: 42
diff of internalized vector vs input vector:
added: input.packet.hex: "ffeeddccbbaa554433221100000f74657374207061636b657420233432"


================================================================================
InternalizeUserProvidedTestVectors Test: single IPv6 packet expected to be forwarded
================================================================================
-- Input ---------------------------------------------------
-- Input Packet Test Vector #1 --
input {
  type: DATAPLANE
  packet {
    port: "1"
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "ff:ee:dd:cc:bb:aa"
          ethernet_source: "55:44:33:22:11:00"
          ethertype: "0x86dd"
        }
      }
      headers {
        ipv6_header {
          version: "0x6"
          dscp: "0x1b"
          ecn: "0x1"
          flow_label: "0x12345"
          next_header: "0xfd"
          hop_limit: "0xf2"
          ipv6_source: "2001::2"
          ipv6_destination: "2001::4"
        }
      }
      payload: "test packet #42"
    }
  }
}
acceptable_outputs {
  packets {
    port: "8"
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "1:2:3:4:5:6"
          ethernet_source: "6:5:4:3:2:1"
          ethertype: "0x86dd"
        }
      }
      headers {
        ipv6_header {
          version: "0x6"
          dscp: "0x1b"
          ecn: "0x1"
          flow_label: "0x12345"
          next_header: "0xfd"
          hop_limit: "0xf1"
          ipv6_source: "2001::2"
          ipv6_destination: "2001::4"
        }
      }
      payload: "test packet #42"
    }
  }
}

-- Output --------------------------------------------------
-- Internalized Packet Test Vector #1 --
test packet ID extracted from payload: 42
diff of internalized vector vs input vector:
added: input.packet.parsed.headers[1].ipv6_header.payload_length: "0x000f"
added: input.packet.hex: "ffeeddccbbaa55443322110086dd66d12345000ffdf2200100000000000000000000000000022001000000000000000000000000000474657374207061636b657420233432"
added: acceptable_outputs[0].packets[0].parsed.headers[1].ipv6_header.payload_length: "0x000f"
added: acceptable_outputs[0].packets[0].hex: "01020304050606050403020186dd66d12345000ffdf1200100000000000000000000000000022001000000000000000000000000000474657374207061636b657420233432"


================================================================================
InternalizeUserProvidedTestVectors Test: several packets with different IDs
================================================================================
-- Input ---------------------------------------------------
-- Input Packet Test Vector #1 --
input {
  type: DATAPLANE
  packet {
    port: "1"
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "42:42:42:42:42:42"
          ethernet_source: "42:42:42:42:42:42"
          ethertype: "0x000f"
        }
      }
      payload: "test packet #42"
    }
  }
}
acceptable_outputs {
}
-- Input Packet Test Vector #2 --
input {
  type: DATAPLANE
  packet {
    port: "1"
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "5:5:5:5:5:5"
          ethernet_source: "5:5:5:5:5:5"
          ethertype: "0x000e"
        }
      }
      payload: "test packet #5"
    }
  }
}
acceptable_outputs {
}

-- Output --------------------------------------------------
-- Internalized Packet Test Vector #1 --
test packet ID extracted from payload: 42
diff of internalized vector vs input vector:
added: input.packet.hex: "424242424242424242424242000f74657374207061636b657420233432"

-- Internalized Packet Test Vector #2 --
test packet ID extracted from payload: 5
diff of internalized vector vs input vector:
added: input.packet.hex: "050505050505050505050505000e74657374207061636b6574202335"


================================================================================
InternalizeUserProvidedTestVectors Test: packet_in with matching tag
================================================================================
-- Input ---------------------------------------------------
-- Input Packet Test Vector #1 --
input {
  type: DATAPLANE
  packet {
    port: "1"
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "ff:ee:dd:cc:bb:aa"
          ethernet_source: "55:44:33:22:11:00"
          ethertype: "0x000f"
        }
      }
      payload: "test packet #42"
    }
  }
}
acceptable_outputs {
  packet_ins {
    metadata {
      name: "ingress_port"
      value {
        hex_str: "0x00f"
      }
    }
    metadata {
      name: "target_egress_port"
      value {
        str: "1"
      }
    }
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "5:5:5:5:5:5"
          ethernet_source: "5:5:5:5:5:5"
          ethertype: "0x000f"
        }
      }
      payload: "test packet #42"
    }
  }
}

-- Output --------------------------------------------------
-- Internalized Packet Test Vector #1 --
test packet ID extracted from payload: 42
diff of internalized vector vs input vector:
added: input.packet.hex: "ffeeddccbbaa554433221100000f74657374207061636b657420233432"
added: acceptable_outputs[0].packet_ins[0].hex: "050505050505050505050505000f74657374207061636b657420233432"


================================================================================
InternalizeUserProvidedTestVectors Test: input type != DATAPLANE
================================================================================
-- Input ---------------------------------------------------
-- Input Packet Test Vector #1 --
input {
  type: SUBMIT_TO_INGRESS
  packet {
    port: "1"
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "ff:ee:dd:cc:bb:aa"
          ethernet_source: "55:44:33:22:11:00"
          ethertype: "0x000f"
        }
      }
      payload: "test packet #42"
    }
  }
}
acceptable_outputs {
}

-- Output --------------------------------------------------
ERROR: UNIMPLEMENTED: problem in user-provided packet test vector: only supported input type is DATAPLANE; found SUBMIT_TO_INGRESS
Dumping offending test vector:
input {
  type: SUBMIT_TO_INGRESS
  packet {
    port: "1"
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "ff:ee:dd:cc:bb:aa"
          ethernet_source: "55:44:33:22:11:00"
          ethertype: "0x000f"
        }
      }
      payload: "test packet #42"
    }
  }
}
acceptable_outputs {
}


================================================================================
InternalizeUserProvidedTestVectors Test: missing expectation
================================================================================
-- Input ---------------------------------------------------
-- Input Packet Test Vector #1 --
input {
  type: DATAPLANE
  packet {
    port: "1"
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "42:42:42:42:42:42"
          ethernet_source: "42:42:42:42:42:42"
          ethertype: "0x000f"
        }
      }
      payload: "test packet #42"
    }
  }
}

-- Output --------------------------------------------------
ERROR: INVALID_ARGUMENT: problem in user-provided packet test vector: must specify at least 1 acceptable output, but 0 were found
Dumping offending test vector:
input {
  type: DATAPLANE
  packet {
    port: "1"
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "42:42:42:42:42:42"
          ethernet_source: "42:42:42:42:42:42"
          ethertype: "0x000f"
        }
      }
      payload: "test packet #42"
    }
  }
}


================================================================================
InternalizeUserProvidedTestVectors Test: missing test packet ID
================================================================================
-- Input ---------------------------------------------------
-- Input Packet Test Vector #1 --
input {
  type: DATAPLANE
  packet {
    port: "1"
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "42:42:42:42:42:42"
          ethernet_source: "42:42:42:42:42:42"
          ethertype: "0x0002"
        }
      }
      payload: "Hi"
    }
  }
}

-- Output --------------------------------------------------
ERROR: INVALID_ARGUMENT: problem in user-provided packet test vector: invalid input packet: test packets must contain a tag of the form 'test packet #([0-9]+)' in their payload, but the given packet with payload 'Hi' does not:
headers {
  ethernet_header {
    ethernet_destination: "42:42:42:42:42:42"
    ethernet_source: "42:42:42:42:42:42"
    ethertype: "0x0002"
  }
}
payload: "Hi"

Dumping offending test vector:
input {
  type: DATAPLANE
  packet {
    port: "1"
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "42:42:42:42:42:42"
          ethernet_source: "42:42:42:42:42:42"
          ethertype: "0x0002"
        }
      }
      payload: "Hi"
    }
  }
}


================================================================================
InternalizeUserProvidedTestVectors Test: inconsistent test packet ID in input vs output
================================================================================
-- Input ---------------------------------------------------
-- Input Packet Test Vector #1 --
input {
  type: DATAPLANE
  packet {
    port: "1"
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "42:42:42:42:42:42"
          ethernet_source: "42:42:42:42:42:42"
          ethertype: "0x000f"
        }
      }
      payload: "test packet #42"
    }
  }
}
acceptable_outputs {
  packets {
    port: "8"
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "5:5:5:5:5:5"
          ethernet_source: "5:5:5:5:5:5"
          ethertype: "0x000f"
        }
      }
      payload: "test packet #24"
    }
  }
}

-- Output --------------------------------------------------
ERROR: INVALID_ARGUMENT: problem in user-provided packet test vector: mismatch of input packet tag vs output packet tag for output packet #1: 42 vs 24
Dumping offending test vector:
input {
  type: DATAPLANE
  packet {
    port: "1"
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "42:42:42:42:42:42"
          ethernet_source: "42:42:42:42:42:42"
          ethertype: "0x000f"
        }
      }
      payload: "test packet #42"
    }
  }
}
acceptable_outputs {
  packets {
    port: "8"
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "5:5:5:5:5:5"
          ethernet_source: "5:5:5:5:5:5"
          ethertype: "0x000f"
        }
      }
      payload: "test packet #24"
    }
  }
}


================================================================================
InternalizeUserProvidedTestVectors Test: several packets with same test packet ID
================================================================================
-- Input ---------------------------------------------------
-- Input Packet Test Vector #1 --
input {
  type: DATAPLANE
  packet {
    port: "1"
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "42:42:42:42:42:42"
          ethernet_source: "42:42:42:42:42:42"
          ethertype: "0x000f"
        }
      }
      payload: "test packet #42"
    }
  }
}
acceptable_outputs {
}
-- Input Packet Test Vector #2 --
input {
  type: DATAPLANE
  packet {
    port: "1"
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "5:5:5:5:5:5"
          ethernet_source: "5:5:5:5:5:5"
          ethertype: "0x000f"
        }
      }
      payload: "test packet #42"
    }
  }
}
acceptable_outputs {
}

-- Output --------------------------------------------------
ERROR: INVALID_ARGUMENT: problem in user-provided packet test vector: user-provided packet test vectors must be tagged with unique IDs in their payload, but found multiple test vectors with ID 42. Dumping offending test vectors:
<input {
  type: DATAPLANE
  packet {
    port: "1"
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "42:42:42:42:42:42"
          ethernet_source: "42:42:42:42:42:42"
          ethertype: "0x000f"
        }
      }
      payload: "test packet #42"
    }
    hex: "424242424242424242424242000f74657374207061636b657420233432"
  }
}
acceptable_outputs {
}
>
<input {
  type: DATAPLANE
  packet {
    port: "1"
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "5:5:5:5:5:5"
          ethernet_source: "5:5:5:5:5:5"
          ethertype: "0x000f"
        }
      }
      payload: "test packet #42"
    }
    hex: "050505050505050505050505000f74657374207061636b657420233432"
  }
}
acceptable_outputs {
}
>

Dumping offending test vector:
input {
  type: DATAPLANE
  packet {
    port: "1"
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "5:5:5:5:5:5"
          ethernet_source: "5:5:5:5:5:5"
          ethertype: "0x000f"
        }
      }
      payload: "test packet #42"
    }
  }
}
acceptable_outputs {
}


================================================================================
InternalizeUserProvidedTestVectors Test: invalid input packet
================================================================================
-- Input ---------------------------------------------------
-- Input Packet Test Vector #1 --
input {
  type: DATAPLANE
  packet {
    port: "1"
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "5:5:5:5:5:5"
          ethernet_source: "5:5:5:5:5:5"
          ethertype: "0x86dd"
        }
      }
      payload: "test packet #42"
    }
  }
}
acceptable_outputs {
}

-- Output --------------------------------------------------
ERROR: INVALID_ARGUMENT: problem in user-provided packet test vector: invalid input packet: Packet invalid for the following reasons:
- in EthernetHeader headers[0]: expected at least 46 bytes of Ethernet payload, but got only 15
- headers[1]: header missing - expected Ipv6Header
Dumping offending test vector:
input {
  type: DATAPLANE
  packet {
    port: "1"
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "5:5:5:5:5:5"
          ethernet_source: "5:5:5:5:5:5"
          ethertype: "0x86dd"
        }
      }
      payload: "test packet #42"
    }
  }
}
acceptable_outputs {
}


================================================================================
InternalizeUserProvidedTestVectors Test: packet_in with mismatching tag
================================================================================
-- Input ---------------------------------------------------
-- Input Packet Test Vector #1 --
input {
  type: DATAPLANE
  packet {
    port: "1"
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "ff:ee:dd:cc:bb:aa"
          ethernet_source: "55:44:33:22:11:00"
          ethertype: "0x000f"
        }
      }
      payload: "test packet #42"
    }
  }
}
acceptable_outputs {
  packet_ins {
    metadata {
      name: "ingress_port"
      value {
        hex_str: "0x00f"
      }
    }
    metadata {
      name: "target_egress_port"
      value {
        str: "1"
      }
    }
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "5:5:5:5:5:5"
          ethernet_source: "5:5:5:5:5:5"
          ethertype: "0x000f"
        }
      }
      payload: "test packet #24"
    }
  }
}

-- Output --------------------------------------------------
ERROR: INVALID_ARGUMENT: problem in user-provided packet test vector: mismatch of input packet in tag vs output packet in tag for output packet in #1: 42 vs 24
Dumping offending test vector:
input {
  type: DATAPLANE
  packet {
    port: "1"
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "ff:ee:dd:cc:bb:aa"
          ethernet_source: "55:44:33:22:11:00"
          ethertype: "0x000f"
        }
      }
      payload: "test packet #42"
    }
  }
}
acceptable_outputs {
  packet_ins {
    metadata {
      name: "ingress_port"
      value {
        hex_str: "0x00f"
      }
    }
    metadata {
      name: "target_egress_port"
      value {
        str: "1"
      }
    }
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "5:5:5:5:5:5"
          ethernet_source: "5:5:5:5:5:5"
          ethertype: "0x000f"
        }
      }
      payload: "test packet #24"
    }
  }
}


================================================================================
InternalizeUserProvidedTestVectors Test: packet_in with malformed metadata
================================================================================
-- Input ---------------------------------------------------
-- Input Packet Test Vector #1 --
input {
  type: DATAPLANE
  packet {
    port: "1"
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "ff:ee:dd:cc:bb:aa"
          ethernet_source: "55:44:33:22:11:00"
          ethertype: "0x000f"
        }
      }
      payload: "test packet #42"
    }
  }
}
acceptable_outputs {
  packet_ins {
    metadata {
      name: "foobar"
      value {
        str: "foobar"
      }
    }
    metadata {
      name: "ingress_port"
      value {
        hex_str: "0x00f"
      }
    }
    metadata {
      name: "target_egress_port"
      value {
        str: "1"
      }
    }
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "5:5:5:5:5:5"
          ethernet_source: "5:5:5:5:5:5"
          ethertype: "0x000f"
        }
      }
      payload: "test packet #24"
    }
  }
}

-- Output --------------------------------------------------
ERROR: INVALID_ARGUMENT: problem in user-provided packet test vector: output packet in #1 invalid: 'packet-in' message is invalid: Metadata with name foobar not defined.
Dumping offending test vector:
input {
  type: DATAPLANE
  packet {
    port: "1"
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "ff:ee:dd:cc:bb:aa"
          ethernet_source: "55:44:33:22:11:00"
          ethertype: "0x000f"
        }
      }
      payload: "test packet #42"
    }
  }
}
acceptable_outputs {
  packet_ins {
    metadata {
      name: "foobar"
      value {
        str: "foobar"
      }
    }
    metadata {
      name: "ingress_port"
      value {
        hex_str: "0x00f"
      }
    }
    metadata {
      name: "target_egress_port"
      value {
        str: "1"
      }
    }
    parsed {
      headers {
        ethernet_header {
          ethernet_destination: "5:5:5:5:5:5"
          ethernet_source: "5:5:5:5:5:5"
          ethertype: "0x000f"
        }
      }
      payload: "test packet #24"
    }
  }
}
